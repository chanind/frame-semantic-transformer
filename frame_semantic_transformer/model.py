from __future__ import annotations
from typing import Any, Optional
import pytorch_lightning as pl
from transformers import (
    T5Tokenizer,
    T5ForConditionalGeneration,
    AdamW,
    get_linear_schedule_with_warmup,
)
import torch
from torch.utils.data import DataLoader, Dataset

from frame_semantic_transformer.data.SentenceDataset import SentenceDataset
from frame_semantic_transformer.data.load_framenet_samples import load_framenet_samples
from frame_semantic_transformer.data.shuffle_and_split import shuffle_and_split


class T5FineTuner(pl.LightningModule):
    """
    Based on https://colab.research.google.com/github/patil-suraj/exploring-T5/blob/master/t5_fine_tuning.ipynb
    """

    lr: float
    eps: float
    weight_decay: float
    batch_size: int
    train_dataset: Dataset[Any]
    val_dataset: Dataset[Any]
    n_gpu: int
    num_train_epochs: int
    warmup_steps: int
    gradient_accumulation_steps: int
    model: T5ForConditionalGeneration
    tokenizer: T5Tokenizer

    def __init__(
        self,
        model: T5ForConditionalGeneration,
        tokenizer: T5Tokenizer,
        lr: float = 1e-4,
        eps: float = 1e-8,
        weight_decay: float = 0.0,
        batch_size: int = 8,
        n_gpu: int = 0,
        num_train_epochs: int = 2,
        gradient_accumulation_steps: int = 16,
        warmup_steps: int = 0,
    ):
        super().__init__()
        self.lr = lr
        self.eps = eps
        self.weight_decay = weight_decay
        self.model = model
        self.tokenizer = tokenizer
        self.batch_size = batch_size
        self.n_gpu = n_gpu
        self.num_train_epochs = num_train_epochs
        self.gradient_accumulation_steps = gradient_accumulation_steps
        self.warmup_steps = warmup_steps

        samples = load_framenet_samples()
        train_samples, val_samples = shuffle_and_split(samples, 0.8)
        self.train_dataset = SentenceDataset(train_samples, tokenizer)
        self.val_dataset = SentenceDataset(val_samples, tokenizer)

    def is_logger(self) -> bool:
        # keep mypy happy...
        if not self.trainer:
            return False
        return self.trainer.global_rank <= 0

    def forward(
        self,
        *args: Any,
        **kwargs: Any,
    ) -> Any:
        return self.model(*args, **kwargs)

    def _step(self, batch: Any) -> Any:
        outputs = self(
            input_ids=batch["source_ids"],
            attention_mask=batch["source_mask"],
            labels=batch["labels"],
        )

        loss = outputs[0]

        return loss

    def training_step(self, batch: Any, batch_idx: int) -> dict[str, Any]:  # type: ignore
        loss = self._step(batch)

        tensorboard_logs = {"train_loss": loss}
        return {"loss": loss, "log": tensorboard_logs}

    def training_epoch_end(self, outputs: Any) -> dict[str, Any]:  # type: ignore
        avg_train_loss = torch.stack([x["loss"] for x in outputs]).mean()
        tensorboard_logs = {"avg_train_loss": avg_train_loss}
        return {
            "avg_train_loss": avg_train_loss,
            "log": tensorboard_logs,
            "progress_bar": tensorboard_logs,
        }

    def validation_step(self, batch: Any, batch_idx: int) -> dict[str, Any]:  # type: ignore
        loss = self._step(batch)
        return {"val_loss": loss}

    def validation_epoch_end(self, outputs) -> dict[str, Any]:  # type: ignore
        avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
        tensorboard_logs = {"val_loss": avg_loss}
        return {
            "avg_val_loss": avg_loss,
            "log": tensorboard_logs,
            "progress_bar": tensorboard_logs,
        }

    def configure_optimizers(self) -> list[AdamW]:
        "Prepare optimizer and schedule (linear warmup and decay)"

        model = self.model
        no_decay = ["bias", "LayerNorm.weight"]
        optimizer_grouped_parameters: Any = [
            {
                "params": [
                    p
                    for n, p in model.named_parameters()
                    if not any(nd in n for nd in no_decay)
                ],
                "weight_decay": self.weight_decay,
            },
            {
                "params": [
                    p
                    for n, p in model.named_parameters()
                    if any(nd in n for nd in no_decay)
                ],
                "weight_decay": 0.0,
            },
        ]
        optimizer = AdamW(
            optimizer_grouped_parameters,
            lr=self.lr,
            eps=self.eps,
        )
        self.opt = optimizer
        return [optimizer]

    def optimizer_step(  # type: ignore
        self,
        epoch: int,
        batch_idx: int,
        optimizer: Any,
        optimizer_idx: int,
        second_order_closure: Optional[Any] = None,
    ) -> None:
        optimizer.step()
        optimizer.zero_grad()
        self.lr_scheduler.step()

    def get_tqdm_dict(self) -> dict[str, Any]:
        tqdm_dict = {
            "loss": "{:.3f}".format(self.trainer.avg_loss),  # type: ignore
            "lr": self.lr_scheduler.get_last_lr()[-1],
        }

        return tqdm_dict

    def train_dataloader(self) -> DataLoader[Any]:
        dataloader: Any = DataLoader(
            self.val_dataset,
            batch_size=self.batch_size,
            num_workers=2,
            shuffle=True,
        )
        t_total = (
            (len(dataloader.dataset) // (self.batch_size * max(1, self.n_gpu)))
            // self.gradient_accumulation_steps
            * float(self.num_train_epochs)
        )
        scheduler = get_linear_schedule_with_warmup(
            self.opt,
            num_warmup_steps=self.warmup_steps,
            num_training_steps=t_total,
        )
        self.lr_scheduler = scheduler
        return dataloader

    def val_dataloader(self) -> DataLoader[Any]:
        return DataLoader(self.val_dataset, batch_size=self.batch_size, num_workers=2)
