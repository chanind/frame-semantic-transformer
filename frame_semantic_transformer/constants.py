MODEL_MAX_LENGTH = 512
OFFICIAL_RELEASES = ["base", "small"]  # TODO: small, large
MODEL_REVISION = "v0.0.1"
