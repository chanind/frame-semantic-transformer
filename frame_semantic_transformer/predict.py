from __future__ import annotations
from typing import Iterable
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration


def predict(
    model: T5ForConditionalGeneration,
    tokenizer: T5Tokenizer,
    source_text: str,
    max_length: int = 512,
    num_return_sequences: int = 1,
    num_beams: int = 5,
    top_k: int = 50,
    top_p: float = 0.95,
    repetition_penalty: float = 2.5,
    length_penalty: float = 1.0,
    early_stopping: bool = True,
    skip_special_tokens: bool = True,
    clean_up_tokenization_spaces: bool = True,
) -> list[str]:
    return batch_predict(
        model,
        tokenizer,
        [source_text],
        max_length=max_length,
        num_return_sequences=num_return_sequences,
        num_beams=num_beams,
        top_k=top_k,
        top_p=top_p,
        repetition_penalty=repetition_penalty,
        length_penalty=length_penalty,
        early_stopping=early_stopping,
        skip_special_tokens=skip_special_tokens,
        clean_up_tokenization_spaces=clean_up_tokenization_spaces,
    )


def batch_predict(
    model: T5ForConditionalGeneration,
    tokenizer: T5Tokenizer,
    source_texts: Iterable[str],
    max_length: int = 512,
    num_return_sequences: int = 1,
    num_beams: int = 5,
    top_k: int = 50,
    top_p: float = 0.95,
    repetition_penalty: float = 2.5,
    length_penalty: float = 1.0,
    early_stopping: bool = True,
    skip_special_tokens: bool = True,
    clean_up_tokenization_spaces: bool = True,
) -> list[str]:
    input_encoding = tokenizer(
        source_texts,
        padding="longest",
        max_length=max_length,
        truncation=True,
        return_tensors="pt",
    )
    return predict_on_ids(
        model,
        tokenizer,
        input_ids=input_encoding.input_ids,
        attention_mask=input_encoding.attention_mask,
        max_length=max_length,
        num_return_sequences=num_return_sequences,
        num_beams=num_beams,
        top_k=top_k,
        top_p=top_p,
        repetition_penalty=repetition_penalty,
        length_penalty=length_penalty,
        early_stopping=early_stopping,
        skip_special_tokens=skip_special_tokens,
        clean_up_tokenization_spaces=clean_up_tokenization_spaces,
    )


def predict_on_ids(
    model: T5ForConditionalGeneration,
    tokenizer: T5Tokenizer,
    input_ids: torch.Tensor,
    attention_mask: torch.Tensor,
    max_length: int = 512,
    num_return_sequences: int = 1,
    num_beams: int = 5,
    top_k: int = 50,
    top_p: float = 0.95,
    repetition_penalty: float = 2.5,
    length_penalty: float = 1.0,
    early_stopping: bool = True,
    skip_special_tokens: bool = True,
    clean_up_tokenization_spaces: bool = True,
) -> list[str]:
    generated_ids = model.generate(
        input_ids=input_ids.to(model.device),
        attention_mask=attention_mask.to(model.device),
        num_beams=num_beams,
        max_length=max_length,
        repetition_penalty=repetition_penalty,
        length_penalty=length_penalty,
        early_stopping=early_stopping,
        top_p=top_p,
        top_k=top_k,
        num_return_sequences=num_return_sequences,
    )
    preds = [
        tokenizer.decode(
            generated_id,
            skip_special_tokens=skip_special_tokens,
            clean_up_tokenization_spaces=clean_up_tokenization_spaces,
        )
        for generated_id in generated_ids
    ]
    return preds
