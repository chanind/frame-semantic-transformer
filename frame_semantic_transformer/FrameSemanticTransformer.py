from __future__ import annotations
from dataclasses import dataclass
from typing import cast
import torch
from transformers import T5ForConditionalGeneration, T5Tokenizer

from frame_semantic_transformer.constants import (
    MODEL_MAX_LENGTH,
    MODEL_REVISION,
    OFFICIAL_RELEASES,
)
from frame_semantic_transformer.data.data_utils import chunk_list, marked_string_to_locs
from frame_semantic_transformer.data.framenet import ensure_framenet_downloaded
from frame_semantic_transformer.predict import batch_predict
from frame_semantic_transformer.data.tasks import (
    ArgumentsExtractionTask,
    FrameClassificationTask,
    TriggerIdentificationTask,
)


@dataclass
class FrameElementResult:
    name: str
    text: str


@dataclass
class FrameResult:
    name: str
    trigger_location: int
    frame_elements: list[FrameElementResult]


@dataclass
class DetectFramesResult:
    sentence: str
    trigger_locations: list[int]
    frames: list[FrameResult]


class FrameSemanticTransformer:
    _model: T5ForConditionalGeneration | None = None
    _tokenizer: T5Tokenizer | None = None
    model_path: str
    model_revision: str | None = None
    device: torch.device
    max_batch_size: int
    predictions_per_sample: int

    def __init__(
        self,
        model_name_or_path: str = "base",
        use_gpu: bool = torch.cuda.is_available(),
        max_batch_size: int = 8,
        predictions_per_sample: int = 5,
    ):
        self.model_path = model_name_or_path
        if model_name_or_path in OFFICIAL_RELEASES:
            self.model_path = f"chanind/frame-semantic-transformer-{model_name_or_path}"
            self.model_revision = MODEL_REVISION
        self.device = torch.device("cuda" if use_gpu else "cpu")
        self.max_batch_size = max_batch_size
        self.predictions_per_sample = predictions_per_sample

    def setup(self) -> None:
        """
        Initialize the model and tokenizer, and download models / files as needed
        If this is not called explicitly it will be lazily called before inference
        """
        self._model = T5ForConditionalGeneration.from_pretrained(
            self.model_path, revision=self.model_revision
        ).to(self.device)
        self._tokenizer = T5Tokenizer.from_pretrained(
            self.model_path,
            revision=self.model_revision,
            model_max_length=MODEL_MAX_LENGTH,
        )
        ensure_framenet_downloaded()

    @property
    def model(self) -> T5ForConditionalGeneration:
        if not self._model:
            self.setup()
        return cast(T5ForConditionalGeneration, self._model)

    @property
    def tokenizer(self) -> T5Tokenizer:
        if not self._tokenizer:
            self.setup()
        return cast(T5Tokenizer, self._tokenizer)

    def _batch_predict(self, inputs: list[str]) -> list[str]:
        """
        helper to avoid needing to repeatedly pass in the same params every call to predict
        """
        return batch_predict(
            self.model,
            self.tokenizer,
            inputs,
            num_beams=self.predictions_per_sample,
            num_return_sequences=self.predictions_per_sample,
        )

    def _identify_triggers(self, sentence: str) -> tuple[str, list[int]]:
        task = TriggerIdentificationTask(text=sentence)
        outputs = self._batch_predict([task.get_input()])
        result = task.parse_output(outputs)
        return marked_string_to_locs(result)

    def _classify_frames(
        self, sentence: str, trigger_locs: list[int]
    ) -> list[str | None]:
        """
        Return a list containing a frame for each trigger_loc passed in.
        If no frame can be found, None is returned for the frame instead.
        """
        frame_classification_tasks: list[FrameClassificationTask] = []
        frames: list[str | None] = []

        for trigger_loc in trigger_locs:
            frame_classification_tasks.append(
                FrameClassificationTask(text=sentence, trigger_loc=trigger_loc)
            )
        for batch in chunk_list(
            frame_classification_tasks, chunk_size=self.max_batch_size
        ):
            batch_results = self._batch_predict([task.get_input() for task in batch])
            for preds, frame_task in zip(
                chunk_list(batch_results, self.predictions_per_sample),
                batch,
            ):
                frames.append(frame_task.parse_output(preds))
        return frames

    def _extract_frame_args(
        self, sentence: str, frame_with_trigger_locs: list[tuple[str, int]]
    ) -> list[list[tuple[str, str]]]:
        """
        return a list of tuples of (frame_element, text) for each frame/trigger loc passed in.
        The returned list will have the same length as the frame_with_trigger_locs list,
        with each element corresponding to a frame/loc in the input list
        """
        frame_element_results: list[list[tuple[str, str]]] = []
        arg_extraction_tasks = [
            ArgumentsExtractionTask(
                text=sentence,
                trigger_loc=trigger_loc,
                frame=frame,
            )
            for frame, trigger_loc in frame_with_trigger_locs
        ]
        for args_tasks_batch in chunk_list(
            arg_extraction_tasks, chunk_size=self.max_batch_size
        ):
            batch_results = self._batch_predict(
                [task.get_input() for task in args_tasks_batch],
            )
            for preds, args_task in zip(
                chunk_list(batch_results, self.predictions_per_sample), args_tasks_batch
            ):
                frame_element_results.append(args_task.parse_output(preds))
        return frame_element_results

    def detect_frames(self, sentence: str) -> DetectFramesResult:
        # first detect trigger locations
        base_sentence, trigger_locs = self._identify_triggers(sentence)
        # next detect frames for each trigger
        frames = self._classify_frames(base_sentence, trigger_locs)

        frame_and_locs = [
            (frame, loc) for frame, loc in zip(frames, trigger_locs) if frame
        ]
        frame_elements_lists = self._extract_frame_args(base_sentence, frame_and_locs)
        frame_results: list[FrameResult] = []
        for ((frame, loc), frame_element_tuples) in zip(
            frame_and_locs, frame_elements_lists
        ):
            frame_elements = [
                FrameElementResult(element, text)
                for element, text in frame_element_tuples
            ]
            frame_results.append(
                FrameResult(
                    name=frame,
                    trigger_location=loc,
                    frame_elements=frame_elements,
                )
            )
        return DetectFramesResult(
            base_sentence,
            trigger_locations=trigger_locs,
            frames=frame_results,
        )
