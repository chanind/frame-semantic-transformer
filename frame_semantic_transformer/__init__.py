__version__ = "0.3.2"

from .FrameSemanticTransformer import (
    FrameSemanticTransformer,
    DetectFramesResult,
    FrameElementResult,
    FrameResult,
)

__all__ = (
    "FrameSemanticTransformer",
    "DetectFramesResult",
    "FrameElementResult",
    "FrameResult",
)
