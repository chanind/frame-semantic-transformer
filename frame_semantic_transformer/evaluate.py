from __future__ import annotations
from collections import defaultdict
from typing import Any, Sequence, Type
from tqdm import tqdm
from transformers import T5ForConditionalGeneration, T5Tokenizer

from frame_semantic_transformer.data.data_utils import chunk_list
from frame_semantic_transformer.data.tasks.ArgumentsExtractionSample import (
    ArgumentsExtractionSample,
)
from frame_semantic_transformer.data.tasks.FrameClassificationSample import (
    FrameClassificationSample,
)
from frame_semantic_transformer.data.tasks.TaskSample import TaskSample
from frame_semantic_transformer.data.tasks.TriggerIdentificationSample import (
    TriggerIdentificationSample,
)
from frame_semantic_transformer.predict import batch_predict, predict_on_ids


def calc_eval_metrics(
    true_pos: float, false_pos: float, false_neg: float
) -> dict[str, float]:
    """
    Calculate precision, recall, and f score
    Based on https://github.com/swabhs/open-sesame/blob/master/sesame/evaluation.py
    """
    if true_pos == 0 and false_pos == 0:
        precision = 0.0
    else:
        precision = true_pos / (true_pos + false_pos)
    if true_pos == 0 and false_neg == 0:
        recall = 0.0
    else:
        recall = true_pos / (true_pos + false_neg)
    if precision == 0 and recall == 0:
        f_score = 0.0
    else:
        f_score = 2.0 * precision * recall / (precision + recall)
    return {"precision": precision, "recall": recall, "f_score": f_score}


def evaluate(
    model: T5ForConditionalGeneration,
    tokenizer: T5Tokenizer,
    samples: Sequence[TaskSample],
    batch_size: int = 10,
    print_failures: bool = False,
    predictions_per_sample: int = 5,
    top_k: int = 50,
    top_p: float = 0.95,
    repetition_penalty: float = 2.5,
    length_penalty: float = 1.0,
    early_stopping: bool = True,
    skip_special_tokens: bool = True,
    clean_up_tokenization_spaces: bool = True,
) -> dict[str, list[float]]:
    results: dict[str, list[float]] = defaultdict(lambda: [0, 0, 0])
    for samples_chunk in tqdm(
        chunk_list(samples, batch_size), total=len(samples) / batch_size
    ):
        inputs = [sample.get_input() for sample in samples_chunk]

        predictions = batch_predict(
            model,
            tokenizer,
            inputs,
            num_beams=predictions_per_sample,
            num_return_sequences=predictions_per_sample,
            top_k=top_k,
            top_p=top_p,
            repetition_penalty=repetition_penalty,
            length_penalty=length_penalty,
            early_stopping=early_stopping,
            skip_special_tokens=skip_special_tokens,
            clean_up_tokenization_spaces=clean_up_tokenization_spaces,
        )
        batched_predictions = chunk_list(predictions, predictions_per_sample)
        for sample, preds in zip(samples_chunk, batched_predictions):
            assert len(preds) == predictions_per_sample
            score = sample.evaluate_prediction(
                preds, sample.get_target(), sample.get_input()
            )
            true_pos, false_pos, false_neg = score
            results[sample.get_task_name()][0] += true_pos
            results[sample.get_task_name()][1] += false_pos
            results[sample.get_task_name()][2] += false_neg
            if print_failures and (false_neg > 0 or false_pos > 0):
                print(score)
                print(sample.get_target())
                print(preds)
                print("\n")

    return results


# TODO: figure out a better way to lookup the class from the string
TASK_SAMPLE_CLASS_MAP: dict[str, Type[TaskSample]] = {
    "args_extraction": ArgumentsExtractionSample,
    "frame_classification": FrameClassificationSample,
    "trigger_identification": TriggerIdentificationSample,
}


def evaluate_batch(
    model: T5ForConditionalGeneration,
    tokenizer: T5Tokenizer,
    batch: Any,
    predictions_per_sample: int = 5,
) -> dict[str, list[float]]:
    predictions = predict_on_ids(
        model,
        tokenizer,
        batch["input_ids"],
        batch["attention_mask"],
        skip_special_tokens=True,
        clean_up_tokenization_spaces=True,
        num_beams=predictions_per_sample,
        num_return_sequences=predictions_per_sample,
    )
    batched_predictions = chunk_list(predictions, predictions_per_sample)
    results: dict[str, list[float]] = defaultdict(lambda: [0, 0, 0])
    for preds, task, label, input_ids in zip(
        batched_predictions, batch["task"], batch["labels"], batch["input_ids"]
    ):
        assert len(preds) == predictions_per_sample
        target_tokens = [tok_id for tok_id in label.tolist() if tok_id != -100]
        input_tokens = [tok_id for tok_id in input_ids.tolist() if tok_id != -100]
        target = tokenizer.decode(
            target_tokens, skip_special_tokens=True, clean_up_tokenization_spaces=True
        )
        input = tokenizer.decode(
            input_tokens, skip_special_tokens=True, clean_up_tokenization_spaces=True
        )
        sample_class = TASK_SAMPLE_CLASS_MAP[task]
        true_pos, false_pos, false_neg = sample_class.evaluate_prediction(
            preds, target, input
        )
        results[task][0] += true_pos
        results[task][1] += false_pos
        results[task][2] += false_neg

    return results
