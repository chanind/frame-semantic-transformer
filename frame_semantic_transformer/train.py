from __future__ import annotations
from collections import defaultdict
import logging
import os
import argparse
from typing import Any, Optional
import numpy as np
import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader, Dataset
from transformers import AdamW, T5ForConditionalGeneration, T5Tokenizer
from pytorch_lightning.callbacks.progress import TQDMProgressBar
from pytorch_lightning.callbacks.early_stopping import EarlyStopping
from pytorch_lightning.callbacks.base import Callback
from frame_semantic_transformer.constants import MODEL_MAX_LENGTH

from frame_semantic_transformer.data.TaskSampleDataset import TaskSampleDataset
from frame_semantic_transformer.data.data_utils import trim_batch
from frame_semantic_transformer.data.load_framenet_samples import (
    load_sesame_train_samples,
    load_sesame_test_samples,
    load_sesame_dev_samples,
)
from frame_semantic_transformer.evaluate import calc_eval_metrics, evaluate_batch

DEFAULT_NUM_WORKERS = os.cpu_count() or 2
logger = logging.getLogger(__name__)


class TrainDataModule(pl.LightningDataModule):
    """
    Based on https://github.com/Shivanandroy/simpleT5/blob/main/simplet5/simplet5.py
    """

    batch_size: int
    train_dataset: Dataset[Any]
    val_dataset: Dataset[Any]
    test_dataset: Optional[Dataset[Any]]
    num_workers: int

    def __init__(
        self,
        train_dataset: Dataset[Any],
        val_dataset: Dataset[Any],
        test_dataset: Optional[Dataset[Any]] = None,
        batch_size: int = 8,
        num_workers: int = DEFAULT_NUM_WORKERS,
    ):
        super().__init__()
        self.batch_size = batch_size
        self.num_workers = num_workers

        self.train_dataset = train_dataset
        self.test_dataset = test_dataset
        self.val_dataset = val_dataset

    def train_dataloader(self) -> DataLoader[Any]:
        dataloader: Any = DataLoader(
            self.train_dataset,
            batch_size=self.batch_size,
            num_workers=self.num_workers,
            shuffle=True,
        )
        return dataloader

    def val_dataloader(self) -> DataLoader[Any]:
        return DataLoader(
            self.val_dataset,
            batch_size=self.batch_size,
            num_workers=self.num_workers,
            shuffle=False,
        )

    def test_dataloader(self) -> DataLoader[Any]:
        dataset = self.test_dataset if self.test_dataset else self.val_dataset
        return DataLoader(
            dataset,
            batch_size=self.batch_size,
            num_workers=self.num_workers,
            shuffle=False,
        )


class TrainingModelWrapper(pl.LightningModule):
    """
    Based on https://github.com/Shivanandroy/simpleT5/blob/main/simplet5/simplet5.py
    """

    lr: float
    model: T5ForConditionalGeneration
    tokenizer: T5Tokenizer
    trainer: pl.Trainer
    output_dir: str
    save_only_last_epoch: bool

    def __init__(
        self,
        model: T5ForConditionalGeneration,
        tokenizer: T5Tokenizer,
        lr: float = 1e-4,
        output_dir: str = "outputs",
        save_only_last_epoch: bool = False,
    ):
        super().__init__()
        self.lr = lr
        self.model = model
        self.tokenizer = tokenizer
        self.output_dir = output_dir
        self.save_only_last_epoch = save_only_last_epoch

    def forward(
        self,
        *args: Any,
        **kwargs: Any,
    ) -> Any:
        return self.model(*args, **kwargs)

    def _step(self, batch: Any) -> Any:
        input_ids, attention_mask, labels = trim_batch(
            batch["input_ids"], batch["attention_mask"], batch["labels"]
        )
        return self(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=labels,
        )

    def training_step(self, batch: Any, _batch_idx: int) -> Any:  # type: ignore
        output = self._step(batch)
        loss = output.loss
        self.log(
            "train_loss",
            loss,
            prog_bar=True,
            logger=True,
            on_epoch=True,
            on_step=True,
            batch_size=len(batch["input_ids"]),
        )
        return loss

    def validation_step(self, batch: Any, _batch_idx: int) -> Any:  # type: ignore
        output = self._step(batch)
        loss = output.loss
        metrics = evaluate_batch(self.model, self.tokenizer, batch)
        self.log(
            "val_loss",
            loss,
            prog_bar=True,
            logger=True,
            on_epoch=True,
            on_step=True,
            batch_size=len(batch["input_ids"]),
        )
        return {"loss": loss, "metrics": metrics}

    def test_step(self, batch: Any, _batch_idx: int) -> Any:  # type: ignore
        output = self._step(batch)
        loss = output.loss
        metrics = evaluate_batch(self.model, self.tokenizer, batch)
        self.log(
            "test_loss",
            loss,
            prog_bar=True,
            logger=True,
            batch_size=len(batch["input_ids"]),
        )
        return {"loss": loss, "metrics": metrics}

    def configure_optimizers(self) -> AdamW:
        return AdamW(self.parameters(), lr=self.lr)

    def training_epoch_end(self, training_step_outputs: list[Any]) -> None:
        """save tokenizer and model on epoch end"""
        self.average_training_loss = np.round(
            torch.mean(torch.stack([x["loss"] for x in training_step_outputs])).item(),
            4,
        )
        path = f"{self.output_dir}/epoch-{self.current_epoch}-train-loss-{str(self.average_training_loss)}-val-loss-{str(self.average_validation_loss)}"
        if (
            not self.save_only_last_epoch
            or self.current_epoch == self.trainer.max_epochs - 1
        ):
            self.tokenizer.save_pretrained(path)
            self.model.save_pretrained(path)

    def validation_epoch_end(self, validation_step_outputs: list[Any]) -> None:
        losses = [out["loss"].cpu() for out in validation_step_outputs]
        self.average_validation_loss = np.round(
            torch.mean(torch.stack(losses)).item(),
            4,
        )

        metrics = merge_metrics([out["metrics"] for out in validation_step_outputs])
        for task_name, counts in metrics.items():
            self.log(f"val_{task_name}_f1", calc_eval_metrics(*counts)["f_score"])

    def test_epoch_end(self, test_step_outputs: list[Any]) -> None:
        metrics = merge_metrics([out["metrics"] for out in test_step_outputs])
        for task_name, counts in metrics.items():
            self.log(f"test_{task_name}_f1", calc_eval_metrics(*counts)["f_score"])


def merge_metrics(metrics: list[dict[str, list[int]]]) -> dict[str, list[int]]:
    merged_metrics: dict[str, list[int]] = defaultdict(lambda: [0, 0, 0])
    for metric in metrics:
        for task_name, counts in metric.items():
            merged_metrics[task_name][0] += counts[0]
            merged_metrics[task_name][1] += counts[1]
            merged_metrics[task_name][2] += counts[2]
    return merged_metrics


def train(
    base_model: str = "t5-base",
    batch_size: int = 8,
    max_epochs: int = 10,
    use_gpu: bool = torch.cuda.is_available(),
    output_dir: str = "outputs",
    early_stopping_patience_epochs: int = 0,  # 0 to disable early stopping feature
    precision: int = 32,
    lr: float = 1e-4,
    num_workers: int = DEFAULT_NUM_WORKERS,
    save_only_last_epoch: bool = False,
    balance_tasks: bool = True,
    max_task_duplication_factor: int = 2,
) -> tuple[T5ForConditionalGeneration, T5Tokenizer]:
    device = torch.device("cuda" if use_gpu else "cpu")
    logger.info("loading base T5 model")
    model = T5ForConditionalGeneration.from_pretrained(base_model).to(device)
    tokenizer = T5Tokenizer.from_pretrained(
        base_model, model_max_length=MODEL_MAX_LENGTH
    )

    logger.info("loading train/test/val datasets")
    train_dataset = TaskSampleDataset(
        load_sesame_train_samples(),
        tokenizer,
        balance_tasks=balance_tasks,
        max_task_duplication_factor=max_task_duplication_factor,
        augment_data=True,
    )
    val_dataset = TaskSampleDataset(
        load_sesame_dev_samples(),
        tokenizer,
        balance_tasks=balance_tasks,
        max_task_duplication_factor=max_task_duplication_factor,
    )
    test_dataset = TaskSampleDataset(
        load_sesame_test_samples(),
        tokenizer,
        balance_tasks=balance_tasks,
        max_task_duplication_factor=max_task_duplication_factor,
    )

    data_module = TrainDataModule(
        train_dataset=train_dataset,
        val_dataset=val_dataset,
        test_dataset=test_dataset,
        batch_size=batch_size,
        num_workers=num_workers,
    )

    model_wrapper = TrainingModelWrapper(
        model,
        tokenizer,
        lr=lr,
        output_dir=output_dir,
        save_only_last_epoch=save_only_last_epoch,
    )

    # add callbacks
    callbacks: list[Callback] = [TQDMProgressBar(refresh_rate=5)]

    if early_stopping_patience_epochs > 0:
        early_stop_callback = EarlyStopping(
            monitor="val_loss",
            min_delta=0.00,
            patience=early_stopping_patience_epochs,
            verbose=True,
            mode="min",
        )
        callbacks.append(early_stop_callback)

    # prepare trainer
    trainer = pl.Trainer(
        callbacks=callbacks,
        max_epochs=max_epochs,
        gpus=1 if use_gpu else 0,
        precision=precision,
        log_every_n_steps=1,
    )

    logger.info("beginning training")

    # fit trainer
    trainer.fit(model_wrapper, data_module)

    return model, tokenizer


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Train a new frame semantic transformer model using FrameNet 1.7"
    )
    parser.add_argument(
        "--base-model",
        default="t5-base",
        help="The HuggingFace T5 model to use as a starting point, default t5-base",
    )
    parser.add_argument("--learning-rate", default=1e-4, type=float)
    parser.add_argument("--use-gpu", action="store_true")
    parser.add_argument("--batch-size", type=int, default=8, help="default 8")
    parser.add_argument("--epochs", type=int, default=10, help="default 10")
    parser.add_argument(
        "--output-dir",
        default="outputs",
        help="dir where output models will be saved after each epoch, default ./outputs",
    )
    args = parser.parse_args()

    logger.setLevel(logging.INFO)

    train(
        base_model=args.base_model,
        batch_size=args.batch_size,
        lr=args.learning_rate,
        use_gpu=args.use_gpu,
        max_epochs=args.epochs,
        output_dir=args.output_dir,
    )
