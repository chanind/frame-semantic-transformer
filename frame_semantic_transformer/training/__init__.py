from .TrainingModelWrapper import TrainingModelWrapper
from .train import train
from .evaluate_model import evaluate_model

__all__ = ["train", "TrainingModelWrapper", "evaluate_model"]
