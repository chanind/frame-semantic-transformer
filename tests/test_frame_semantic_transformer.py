from syrupy.assertion import SnapshotAssertion

from frame_semantic_transformer import FrameSemanticTransformer


def test_basic_detect_frames_functionality(snapshot: SnapshotAssertion) -> None:
    transformer = FrameSemanticTransformer("small")
    result = transformer.detect_frames(
        "I'm getting quite hungry, but I can wait a bit longer."
    )
    assert result.sentence == "I'm getting quite hungry, but I can wait a bit longer."
    assert result == snapshot


def test_basic_detect_frames_bulk() -> None:
    sentences = [
        "I'm getting quite hungry, but I can wait a bit longer.",
        "The chef gave the food to the customer.",
        "The hallway smelt of boiled cabbage and old rag mats.",
    ]
    transformer = FrameSemanticTransformer("small")
    results = transformer.detect_frames_bulk(sentences)
    assert len(results) == 3
    for result, sentence in zip(results, sentences):
        assert result == transformer.detect_frames(sentence)
        assert result.sentence == sentence
        assert len(result.trigger_locations) > 0
        assert len(result.frames) > 0
