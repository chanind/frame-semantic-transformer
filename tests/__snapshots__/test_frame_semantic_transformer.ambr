# name: test_basic_detect_frames_functionality
  DetectFramesResult(sentence="I'm getting quite hungry, but I can wait a bit longer.", trigger_locations=[4, 18, 36], frames=[FrameResult(name='Getting', trigger_location=4, frame_elements=[FrameElementResult(name='Recipient', text='I'), FrameElementResult(name='Theme', text='quite hungry')]), FrameResult(name='Biological_urge', trigger_location=18, frame_elements=[FrameElementResult(name='Experiencer', text='I'), FrameElementResult(name='Degree', text='quite')]), FrameResult(name='Waiting', trigger_location=36, frame_elements=[FrameElementResult(name='Protagonist', text='I'), FrameElementResult(name='Time', text='a bit longer')])])
# ---
